Load and coerce data using schema

Eleonore9 opened the pull request:
**Aim**: I need to be able to load data with the right data type in each column, hence data coercion with schema

**Issue**: I takes a long time to run and fail -> issue #9

**Investigation**:
* Researched and schema validation seems to block -> Improve schema coercion
* Coercion faster, but dataset creation fails `ExceptionInfo Can't create dataset from incomplete maps` -> It must me due to missing data in a few rows -> addresses issue #5 

**Solution**:
* Improved schema validation
* Added option to handle errors due to inconsistent data

*Notes*: 
* Running `csv-to-dataset` on the repair data is still long
* The relevance of the second arity in `csv-to-dataset` can be discuss

Commit:
Improve schema validation, still issue due to data

Implement schema coercion changes in test ns.

Commit:
WIP: improve 'load-csv' and handle coercion errors

Commit:
Report and filter out errors in the data.

Commit:
Rename fns, add docstrings and tests.

Commit:
Changes due to lein kibit and lein cljmt fix

Commit:
Doc: update Readme + add tests

acron0 commented on src/kixi/lapita.clj:
```diff
@@ -20,38 +20,57 @@
+     (ds/dataset coerced-data)))
+  ;; Specify options in the 3rd arg to print number of rows with inconsistent data,
+  ;; write inconsistent data to file return only coerced data.
+  ([filename schema {:keys [print-errors write-errors remove-errors]}]
```
When you have an arity with optional additional 'options' it's usual that the version without (two-arity in this case) calls the version with (three-arity in this case), just with an empty options map:

```
(defn foo
  ([x y opts]
    ...)
  ([x y]
    (foo x y {}))
```

Eleonore9 replied:
Oh yes, that's true! Thanks!

acron0 commented on src/kixi/lapita.clj:
```diff
@@ -20,38 +20,57 @@
+     (when print-errors (println (format "There are %d rows with data coercion errors out of %d rows"
+                                         (count errors) (count coerced-data))))
+     (when write-errors
+       (write-csv! errors "data/records-with-data-coercion-issues.csv")
```
Either save in a temp (`/tmp') file or allow the user to specify a location.

acron0 commented on src/kixi/lapita.clj:
```diff
@@ -109,13 +128,24 @@
 
 ;; Examples of uses
-(comment (def repairs-data (load-csv "data/historic-repairs-2014-2015.csv"))
+(comment (def repairs-data (csv-to-maps-coll "data/historic-repairs-2014-2015.csv"))
```
Do these still need to be here? Are they in tests now?

Eleonore9 replied:
Well, they're very handing for me. But I guess I should save those examples locally!

Commit:
Implement @acron0's comments.

Improve `csv-to-dataset`, replace use of metadata, remove dataset specific info

acron0 commented on src/kixi/lapita.clj:
```diff
@@ -53,19 +53,18 @@
   ;; write inconsistent data to file return only coerced data.
-  ([filename schema {:keys [print-errors write-errors remove-errors]}]
+  ([filename schema {:keys [print-errors write-errors errors-filepath remove-errors]
+                     :or {errors-filepath "test-data/records-with-data-coercion-issues.csv"}}]
```
This is still not a safe file path to assume. You either need to force them to specify a dir or create a temporary file using a safe method.

acron0 commented on src/kixi/lapita.clj:
```diff
@@ -53,19 +53,18 @@
      (when write-errors
-       (write-csv! errors "data/records-with-data-coercion-issues.csv")
+       (write-csv! errors errors-filepath)
        (println "The records with data coercion issues were saved in 'data/records-with-data-coercion-issues.csv'"))
```
Output should use `errors-filepath`

acron0 commented on src/kixi/lapita/schemas.clj:
```diff
@@ -1,39 +1,49 @@
+    (pmap (fn [d] (let [coerced (data-coercer d)]
+                    (if (su/error? coerced)
+                      (with-meta d {:data-coercion :errors})
+                      (with-meta coerced {:data-coercion :validated}))))
```
Could we avoid using metadata here and just return a map with `:results` and `:errors`?

acron0 commented on src/kixi/lapita.clj:
```diff
@@ -111,7 +124,7 @@
 (comment (def repairs-data (load-csv "data/historic-repairs-2014-2015.csv"))
```
Remove these pls

acron0 commented on src/kixi/lapita/schemas.clj:
```diff
@@ -1,13 +1,18 @@
   [data schema]
   (let [data-coercer (coerce/coercer schema coerce/string-coercion-matcher)]
-    (pmap data-coercer data)))
+    (pmap (fn [d] (let [coerced (data-coercer d)]
```
I don't think you need this pmap at all now

acron0 commented on src/kixi/lapita/schemas.clj:
```diff
@@ -1,39 +1,52 @@
-        (:column-names col-schema)))
+;; NOTE: schema-contrib.core/Date expect a date in format ISO 8601 like YYYY-MM-DD
+;; To solve that, change the format on the spreadsheet to match ISO 8601.
+(def HousingRepairSchema
```
Move to tests

Eleonore9 replied:
I'm not using it in the test. Would it be a good place for my schema?

acron0 commented on src/kixi/lapita.clj:
```diff
@@ -20,38 +20,56 @@
+   (csv-to-dataset filename schema {}))
+  ;; Specify options in the 3rd arg to print number of rows with inconsistent data,
+  ;; write inconsistent data to file return only coerced data.
+  ([filename schema {:keys [print-errors write-errors errors-filepath remove-errors]}]
```
I wouldn't bother having `write-errors` and `errors-filepath`, just have `write-errors` and let it either be a path to write to or nil.

acron0 reviewed (approved):
looking good!
